packages/plugin-legacy/index.js

```diff
@@ -15,8 +15,20 @@ const loadBabel = () => babel || (babel = require('@babel/standalone'))
 // DO NOT ALTER THIS CONTENT
 const safari10NoModuleFix = `!function(){var e=document,t=e.createElement("script");if(!("noModule"in t)&&"onbeforeload"in t){var n=!1;e.addEventListener("beforeload",(function(e){if(e.target===t)n=!0;else if(!e.target.hasAttribute("nomodule")||!n)return;e.preventDefault()}),!0),t.type="module",t.src=".",e.head.appendChild(t),t.remove()}}();`
 
+const legacyPolyfillId = 'vite-legacy-polyfill'
 const legacyEntryId = 'vite-legacy-entry'
 const systemJSInlineCode = `System.import(document.getElementById('${legacyEntryId}').getAttribute('data-src'))`
+const dynamicFallbackInlineCode = `try {
+  new Function('m', 'return import(m)');
+} catch (e) {
+  var polyfillScript = document.getElementById('${legacyPolyfillId}');
+  var polyfillCopy = document.createElement('script');
+  polyfillCopy.src = polyfillScript.src;
+  polyfillCopy.onload = function() {
+    ${systemJSInlineCode};
+  };
+  document.body.appendChild(polyfillCopy);
+}`
```

Comment:
I believe there's no standard way of stopping script execution from another script tag without cooperation from the tag being executed.
In this example second tag still runs, even though first one throws:

```html
<!DOCTYPE html>
<html>
  <head></head>
  <body>
    <script type="module">
      throw new Error("oops");
    </script>
    <script type="module">
      console.log("still executed");
    </script>
  </body>
</html>
```

I guess one solution might be to inject blank dynamic import to the modern bundle. (i.e. `import('data:text/javascript;base64,Cg==');`

This would force syntax error and prevent execution. Later we can explain what's happening somewhat like this:

```js
console.warn('vite: loading legacy build because dynamic import is unsupported, syntax error above should be ignored')
```

Reply:
But you need the dynamic import in each bundle anyways, to point to its legacy version. You would already stop execution of all of them. Or am I missing something? But yes, each modern bundle should load its legacy counterpart and stop executing.

It is ok also to retwrow the import error, and warn as you said. Whatever is less noisy.

Reply:
I assumed vite produced *exactly* one modern entry point (and one legacy, if the plugin is enabled), maybe that's wrong and then my method is simply incorrect.

The problem I was trying to describe is if the entire bundle didn't have a single `import()` call and therefore would just continue executing. But maybe that a non-issue since this option shouldn't be used in that case anyway.

If my assumption was correct, there shouldn't be any need to include fallback logic inside the modern bundle. Instead, we could have 3 separate scripts: modern bundle, fallback logic and legacy bundle.

- In `nomodule` browsers only the third one would be downloaded and executed.
- In `module` w/o `import()` browsers modern bundle should error out, the legacy bundle would not be downloaded by default, while fallback code would check, re-create, and download the legacy bundle (and system-js runtime).
- In `import()` browsers modern bundle would execute, while the fallback code would just check and do nothing.

Reply:
> I assumed vite produced _exactly_ one modern entry point (and one legacy, if the plugin is enabled), maybe that's wrong and then my method is simply incorrect.

Sorry, you are right. I was thinking about the vendor chunk, but that is only a preload link.

> The problem I was trying to describe is if the entire bundle didn't have a single `import()` call and therefore would just continue executing. But maybe that a non-issue since this option shouldn't be used in that case anyway.

The issue with this is that if there isn't any `import()` call and the browser supports module but no dynamic import, then the bundle will properly execute and the fallback logic will also do so... And even if it doesn't, our targets for esbuild that generated this modern build are not right for these browsers. So we need to avoid executing the modern build even for these cases.

> If my assumption was correct, there shouldn't be any need to include fallback logic inside the modern bundle. Instead, we could have 3 separate scripts: modern bundle, fallback logic and legacy bundle.

I think both approaches would work:

1. modern bundle (with simple import guard that throws), fallback logic module, legacy bundle
2. modern bundle (with import guard that implements fallback logic and then throws), legacy bundle 

Do you see any benefits of one over the other?

PS: I see now that I misunderstood your previous comment
> guess one solution might be to inject blank dynamic import to the modern bundle. (i.e. `import('data:text/javascript;base64,Cg==');` )

This is an option instead of eval-guard+throw if 1. is used. The guard to control the error message looks better to me as it may be easier to ignore than a real syntax error (maybe filter it in some environments, etc). What do you think?